# Scenario

An electronic retailer Electronics online (EO) wants to understand whether the company performed well and how accurate the forecasts have been based on historical data. 

The board will ask what predictions can be made about the stock price given the new inputs. This requires a model that outputs continuous variables. Therefore, we need a regression decision tree model and we can test the model's accuracy after building it.

In [1]:
# Import the libraries
import pandas as pd
import numpy as np
import scipy as scp
import sklearn
import statsmodels.api as sm

from imblearn.over_sampling import SMOTE
from sklearn.model_selection import train_test_split

import warnings
warnings.filterwarnings('ignore')

# Read the csv file 
data = pd.read_csv(r'C:\Users\patri\Desktop\LSE_CareerAccelerator\Course 3\Week 2\Data\ecommerce.csv')

# Print a summary of the DataFrame to sense-check it
data.info()

C:\Users\patri\anaconda3\lib\site-packages\statsmodels\compat\pandas.py:61: FutureWarning: pandas.Int64Index is deprecated and will be removed from pandas in a future version. Use pandas.Index with the appropriate dtype instead.
  from pandas import Int64Index as NumericIndex


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 12 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Sale       506 non-null    float64
 1   por_OS     506 non-null    float64
 2   por_NON    506 non-null    float64
 3   recc       506 non-null    int64  
 4   avg_no_it  506 non-null    float64
 5   age        506 non-null    float64
 6   dis        506 non-null    float64
 7   diff_reg   506 non-null    int64  
 8   tax        506 non-null    int64  
 9   bk         506 non-null    float64
 10  lowstat    506 non-null    float64
 11  Median_s   506 non-null    float64
dtypes: float64(9), int64(3)
memory usage: 47.6 KB


In [2]:
# We can now build the regression decision tree model

# Specify the column Median s will be moved to another dataframe
cols = data.columns[data.columns != 'Median_s']

# Specify 'X' as the independent variables and y as the dependent variable
X = data[cols]
y = data['Median_s']

# Split the data to training and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 0)

# Import the 'DecisionTreeRegressor' class from sklearn
from sklearn.tree import DecisionTreeRegressor

# Create the 'DecisionTreeRegressor' class
regressor = DecisionTreeRegressor(random_state = 0)

# Fit the regressor object to the data set
regressor.fit(X_train, y_train)

DecisionTreeRegressor(random_state=0)

We managed to fit the model and we can now test the accuracy of the model

## Determine the accuracy of the model

This works slightly different from the other approaches BLR, MLR, SVR, Classification Decision Trees... 
We need to calculate

* Mean Absolute Error: this indicates on average how large of an error we expect between the predicted and actual output
* Mean Squared Error: This is the squared error between the differences of predicted vs actual values
* Root  Mean Squared Error: this takes the Mean Squared Error and takes the square root from it. We can compare RMSE and AER to determine whether forecasts contain large but infrequent error. 

The larger the difference the more inconsistent the error size. Therefore we want the difference to be close to 0 as possible



In [7]:
# Import the packages
from sklearn import metrics
import math

# Predict the response for the data test
y_pred = regressor.predict(X_test)

# Specify to print the MAE and MSE to evaluate the accuracy
print('Mean Absolute Error: ', round(metrics.mean_absolute_error(y_test, y_pred), 2))
print('Mean Squared Error: ', round(metrics.mean_squared_error(y_test, y_pred), 2))
print('Root Mean Squared Error: ', round(math.sqrt(metrics.mean_squared_error(y_test, y_pred)),2))

Mean Absolute Error:  3.26
Mean Squared Error:  26.4
Root Mean Squared Error:  5.14


In [9]:
round(metrics.mean_absolute_error(y_test, y_pred), 2) - round(math.sqrt(metrics.mean_squared_error(y_test, y_pred)),2)

-1.88

### Interpretation

The output indicates that the MAE is 3.26 and the MSE is 26.4. The difference between the MAE and RMSE is 1.88 which is considered small. 